src/shims/ffi_support.rs

```diff
@@ -0,0 +1,285 @@
+use libffi::{high::call::*, low::CodePtr};
+use std::ops::Deref;
+
+use rustc_middle::ty::{IntTy, Ty, TyKind, UintTy};
+use rustc_span::Symbol;
+use rustc_target::abi::HasDataLayout;
+
+use crate::*;
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    /// Extract the scalar value from the result of reading a scalar from the machine,
+    /// and convert it to a `CArg`.
+    fn scalar_to_carg(
+        k: ScalarMaybeUninit<Provenance>,
+        arg_type: &Ty<'tcx>,
+        cx: &impl HasDataLayout,
+    ) -> InterpResult<'tcx, CArg> {
+        match arg_type.kind() {
+            // If the primitive provided can be converted to a type matching the type pattern
+            // then create a `CArg` of this primitive value with the corresponding `CArg` constructor.
+            // the ints
+            TyKind::Int(IntTy::I8) => {
+                return Ok(CArg::Int8(k.to_i8()?));
+            }
+            TyKind::Int(IntTy::I16) => {
+                return Ok(CArg::Int16(k.to_i16()?));
+            }
+            TyKind::Int(IntTy::I32) => {
+                return Ok(CArg::Int32(k.to_i32()?));
+            }
+            TyKind::Int(IntTy::I64) => {
+                return Ok(CArg::Int64(k.to_i64()?));
+            }
+            TyKind::Int(IntTy::Isize) => {
+                return Ok(CArg::ISize(k.to_machine_isize(cx)?.try_into().unwrap()));
```

Comment:
```suggestion
                // This will fail if host != target, but then the entire FFI thing probably won't work well
                // in that situation.
                return Ok(CArg::ISize(k.to_machine_isize(cx)?.try_into().unwrap()));
```
Same for the usize case.